presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,314 @@
+    {
+        this(dbi,
+                nodeManager.getCurrentNode().getNodeIdentifier(),
+                shardManager,
+                organizer,
+                config.isOrganizationEnabled(),
+                config.getOrganizationInterval());
+    }
+
+    public ShardOrganizationManager(
+            IDBI dbi,
+            String currentNodeIdentifier,
+            ShardManager shardManager,
+            ShardOrganizer organizer,
+            boolean enabled,
+            Duration organizationInterval)
+    {
+        this.dbi = requireNonNull(dbi, "dbi is null");
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardOrganizerDao = onDemandDao(dbi, ShardOrganizerDao.class);
+
+        this.organizer = requireNonNull(organizer, "organizer is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.currentNodeIdentifier = requireNonNull(currentNodeIdentifier, "currentNodeIdentifier is null");
+
+        this.enabled = enabled;
+
+        requireNonNull(organizationInterval, "organizationInterval is null");
+        this.organizationIntervalSeconds = max(1, organizationInterval.roundTo(SECONDS));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!enabled) {
+            return;
+        }
+
+        if (!started.getAndSet(true)) {
+            startDiscovery();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdownNow();
+    }
+
+    private void startDiscovery()
+    {
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                SECONDS.sleep(ThreadLocalRandom.current().nextLong(1, 5 * 60));
+
+                log.info("Running shard organizer...");
+                Set<Long> tableIds = discoverAndInitializeTablesToOrganize();
+                submitJobs(tableIds);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error running shard organizer");
+            }
+        }, 0, 5, MINUTES);
+    }
+
+    @VisibleForTesting
+    Set<Long> discoverAndInitializeTablesToOrganize()
+    {
+        Set<Long> enabledTableIds = metadataDao.getOrganizationEligibleTables();
+
+        Set<TableOrganizationInfo> tableOrganizationInfo = shardOrganizerDao.getNodeTableOrganizationInfo(currentNodeIdentifier);
+        Map<Long, TableOrganizationInfo> organizationInfos = Maps.uniqueIndex(tableOrganizationInfo, TableOrganizationInfo::getTableId);
+
+        // If this is the first time organizing a table, initialize the organization info for it
+        Set<Long> newTableIds = difference(enabledTableIds, organizationInfos.keySet());
```

Comment:
Inline this variable since it's a view. You don't want someone to accidentally use it thinking that it's a set